v0/destinations/yahoo_dsp/util.js

```diff
@@ -0,0 +1,173 @@
+    }
+  });
+  return includes;
+};
+
+// updating excludes
+const populateExcludes = (list, audienceType) => {
+  list.forEach(element => {
+    const keys = Object.keys(element);
+    keys.forEach(elementKey => {
+      if (poiLocationType.includes(elementKey)) {
+        listType = audienceType;
+      }
+    });
+    if (!listType) {
+      throw new CustomError(`Audience type ${audienceType} not provided`, 400);
+    }
+    if (element.excludeChains) {
+      if (!excludes.chains) {
+        excludes.chains = [];
+      }
+      excludes.chains.push(element.excludeChains);
+    }
+    if (element.excludeWoeids) {
+      if (!excludes.woeids) {
+        excludes.woeids = [];
+      }
+      excludes.woeids.push(element.excludeWoeids);
+    }
+    if (element.excludeGids) {
+      if (!excludes.gids) {
+        excludes.gids = [];
+      }
+      excludes.gids.push(element.excludeGids);
+    }
+    if (element.excludeCategories) {
+      if (!excludes.categories) {
+        excludes.categories = [];
+      }
+      excludes.categories.push(element.excludeCategories);
+    }
+  });
+  return excludes;
+};
+
+/**
+ * The funciton here is used to generate acccess token which needs some parameters like clientId, clientSecret which is being
+ * taken from destination.Config. 
+ * @param {*} destination 
+ * @returns 
+ */
+const getAccessToken = async destination => {
+  const {clientId, clientSecret} = destination.Config;
+  const accessTokenKey = destination.ID;
+
+  /**
+   * The access token expires around every one hour. Cache is used here to check if the access token is present in the cache
+   * it is taken from cache else a post call is made to get the access token.
+   */
+  return accessTokenCache.get(accessTokenKey, async () => {
+    const header = {
+      alg: "HS256",
+      typ: "JWT"
+    };
+
+    const data = {
+      aud: "https://id.b2b.yahooinc.com/identity/oauth2/access_token?realm=dsp",
+      sub: clientId,
+      iss: clientId,
+      exp: getUnixTimestamp(),
+      iat: getUnixTimestamp() + 3600
+    };
+    const secret = clientSecret;
+
+    const request = {
+      header: {
+        "Content-Type": "application/x-www-form-urlencoded",
+        Accept: "application/json"
+      },
+      url: "https://id.b2b.yahooinc.com/identity/oauth2/access_token",
+      data: qs.stringify({
+        grant_type: "client_credentials",
+        scope: "dsp-api-access",
+        realm: "dsp",
+        client_assertion_type:
+          "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+           // Here, generateJWTToken is used to get JWT required for genrating access token.
+        client_assertion: generateJWTToken(header, data, secret)
+      }),
+      method: "POST"
+    };
+    const response = await httpPOST(request.url, request.data, request.header);
+    // If the request fails, throwing error.
+    if(response.success === false){
+      throw new CustomError(`${response.response.response.data.error}`,400);
```

Comment:
Use optional chaining and add some error messages in front. Something like "Unable to generate accessToken", then the error.

Reply:
Updated